esphome/components/api/api_server.cpp

```diff
@@ -399,6 +403,26 @@ void APIServer::send_homeassistant_action(const HomeassistantActionRequest &call
     client->send_homeassistant_action(call);
   }
 }
+
+void APIServer::register_action_response_callback(uint32_t call_id, ActionResponseCallback callback) {
+  this->action_response_callbacks_[call_id] = std::move(callback);
+}
+
+void APIServer::handle_action_response(uint32_t call_id, bool success, const std::string &error_message,
+                                       const std::string &response_data) {
+  auto it = this->action_response_callbacks_.find(call_id);
+  if (it != this->action_response_callbacks_.end()) {
+    // Create the response object
+    auto response = std::make_shared<class ActionResponse>(success, error_message);
+    response->set_data(response_data);
+
+    // Call the callback
+    it->second(response);
+
+    // Remove the callback as it's one-time use
+    this->action_response_callbacks_.erase(it);
```

Comment:
```suggestion
    auto callback = std::move(it->second);
    this->action_response_callbacks_.erase(it);
    auto response = std::make_shared<ActionResponse>(success, error_message, response_data);
    callback(response);
```

Maybe avoid a string copy

Reply:
But you need
```
  ActionResponse(bool success, std::string error_message, std::string data)
      : success_(success), error_message_(std::move(error_message)), data_(std::move(data)) {}
```